### ISSUE TITLE
Battle-related bugs & missing functionalities

### ISSUE THREAD

**Issue by IvanSavenko** [vcmi/vcmi#1219]
This is full list of known issues related to battle functionality that do not have separate issue. 
For others, see [unresolved battle issues on github](https://github.com/vcmi/vcmi/issues?q=is%3Aopen+label%3Abattle+no%3Aassignee+-label%3Aenhancement+-label%3Astupidai)

**General issues:**
- [ ] some obstacles have incorrect rendering order. obstacles.json has "unknown" field which may have "render this obstacle in background" meaning
- [ ] stack amount boxes positions are not updated AFTER move but only at start of next move
- [ ] Hovering over hero should show hero cursor
- [ ] On hex overlapping with hero image interaction with hero should be prioritized
- [ ] Recheck timings of all battle animations
- [ ] Stack amount boxes flicker a bit between stack animations
- [ ] Stack amount boxes update count & spell effect color before actual hit
- [ ] Destination point of spell projectiles seems to be off (at least - ice bolt)
- [ ] Hero animation for obstacle-creating spells plays only after obstacles are placed

**Spell-specific issues:**
- [ ] "Clone": position of created clone is different from H3
- [ ] "Sacrifice": wrong visual effect on sacrificed creature
- [ ] "Sacrifice": animation on resurrected creature should play after effect on sacrificed
- [ ] "Chain lighting": effect should be chained for 2-5th stacks. Probably using same effect as Mage/Evil Eye rays
- [ ] "Remove obstacle": no fade-out effect for removed object
- [ ] "Demon summon": stack animation should be paused for the duration of fade-in animation
- [ ] "Landmine": Visual explosion of mines is missing

**Ability-specific issues:**
- [ ] Lich Death Cloud (and likely Magog Fireball): hit animation on creatures plays BEFORE cloud animation. 
- [ ] Spell resistance (dwarves): effect should only play on primary target. Secondary targets on successful resist take no damage, but play no animations at all - no hit animation, no resistance sound, no resistance visual effect (NOTE: perhaps this can be considered bug in H3? Having Dwarves miraculously survive unharmed after Armageddon looks weird)
- [ ] Spell resistance (dwarves): If primary target resisted the spell, then spell animations should be skipped completely - no sound, no projectile, no hit effects). Hero's casting animation still plays though.

**Potential enhancements**
- [ ] Add better hex highlighting - use custom images for movement path, enemy movement area, spell target area, creature movement destination.
- [ ] Highlight all targeted creatures during dragon breath/all-around attacks

**MANTIS: Missing functionality (UI)**
- [ ] 0002205: Creatures info can not be changed in battle options
- [ ] 0002205: auto-combat rules can not be changed in battle options
- [ ] 0001565: Autobattle should be cancelable by mouse click
- [ ] 0002869: There is no warning if surrendering hero will trigger scenario loss
- [ ] 0000697: Hero info mini-card should appear when hovering over hero
- [ ] 0001408: Missing hotkey: T should open creature window of a hoveres stack
- [ ] 0001408: Missing hotkey: F5 should toggle displayed info on creature card popup
- [ ] 0001408: Missing hotkey: F6: should toggle hex grid
- [ ] 0001408: Missing hotkey: F7: should toggle mouse shadow display
- [ ] 0001408: Missing hotkey: F8: should toggle movement shadow display
- [ ] 0001408: Missing hotkey: S: during tactics phase, should start combat
- [ ] 0001285: On starting combat, if hero has tactics there should be popup "hero with superior tactics may arrange his armies"
- [ ] 0001285: Right-clicking "Next creature" button should show "Phase wait" message
- [ ] 0001285: Right-clicking "Start Combat " button should show "End Tactics Phase" message

**MANTIS: Missing functionality (mechanics)**
- [ ] 0002572: War machines should affect surrender cost
- [ ] 0002676: Zombies should cast Disease after attack with 20% chance
- [ ] 0002676: Bone/Ghost Dragon should have ability "-1 to enemy morale"
- [ ] 0002786: Heroes fighting at the sea should fight on ship-to-ship battle with custom obstacles
- [ ] 0002848: Elementals don't deal 100% higher damage to the elementals of opposing element (i.e. Fire and Water).
- [ ] 0002916: Magic Elementals should only deal 50% damage to all units resistant to 5th level spells
- [ ] 0002915: Psychic Elementals should deal 50% damage to all units resistant to mind spells, including Undead and Non-living creatures (BUT NOT GARGOYLES!). 

**MANTIS: Known bugs (UI):**
- [ ] 0002322: CBattleInterface destructor should wait "activeStack" thread to finish. 
- [ ] 0002392: Starting combat while holding left mouse button may lead to automatic movement of a stack
- [ ] 0000651: When right-clicking a creature, mouse cursor should change to pointer
- [ ] 0000694: Hoverting over corner of spell book shows "Next Page" tooltip even on last page
- [ ] 0002162: Hotseat: popup message "Spell will affect no one" don't appear for 2nd player
- [ ] 0002290: Spells like Disguise / Visions affect color of stack amount boxes
- [ ] 0002240: Actual damage for ranged units is higher than upper damage limit in combat log hint.
- [ ] 0002744: Bloodlust doesn't update the Attack value of the affected creatures 

**MANTIS: Known bugs (mechanics):**
- [ ] 0002893: Elixir of Life should affect Pit Lord's Demon Summoning
- [ ] 0002894: Adela's Bless boosts only max damage, but should affect min damage as well
- [ ] 0002373: Troops lost by retreat & surrender should not be used for necromancy skill
- [ ] 0002770: Killing Dendroid Soldier should remove entangle from affected stacks
- [ ] 0002896: Heroes with specialties connected with stat-altering spells don't do anything.(SPECIAL_PECULIAR_ENCHANT)
- [ ] 0002947: Fights inside mines should always use obstacles from underground terrain
- [ ] 0002971: Current Health doesn't update with Maximum Health if it's lowered e.g. with Poison ability
- [ ] 0003199: Magic elemental should have immunity against Thunderbold (from Thunderbirds)
- [ ] 0001616: It is possible to receive morale after defeating last enemy stack
- [ ] 0002451: Drawbridge stucks in locked state after first creature passes through it during tactics phase
- [ ] 0003165: Creatures like Archangels can cast their spell on self when in auto-combat, which should be impossible
- [ ] 0002807: Pit lord's summoning demons ability should be blocked if 0 creatures can be raised
- [ ] 0003132: creatures affected by "paralyse" and "stoneGaze" spells, should retaliate upon being attacked.

**MANTIS: Unconfirmed bugs:**
- [ ] 0002787: Magic Resistance animation is missing
- [ ] 0002787: Magic Resistance works incorrectly with Chain Lighting
- [ ] 0002614: Ancient Behemoths under Bless spell deal different amount of damage in H3 & vcmi
- [ ] 0002733: Check targeting of area-of-effect spells of Faerie Dragons
- [ ] 0003106: The firewall should hit double-wide units only once when moving through it
- [ ] 0001791: Orb of vulnerability mechanics in vcmi is different from H3
- [ ] 0003052: Adventure map is not deactivated during battles initiated by AI
- [ ] 0000684: Curse spell decreases damage to 1
- [ ] 0001065: Catapult target selection areas are different from H3
- [ ] 0002925: Blind - unit receives turn, but can't do anything

**MANTIS: Enhancements:**
- [ ] 0000950: Allow shooters to move and attack enemy units in close combat 
- [ ] 0000949: Show expected damage, kills and potential counter-attack when mouse is over enemy unit stack
- [ ] 0002417: Add better formula for merging garrison & visitor armies during siege

**Comment by Nordsoft91** [comment]
Wow, that's very comprehensive list and great knowledge of original game!

**Comment by dydzio0614** [comment]
i would add one mor to the list - there should be different dragon breath animation for 1 hex and 2 hex breath and vcmi uses only one of them AFAIK

**Comment by dydzio0614** [comment]
Just reminded myself of one more thing: faerie dragon spell pick should be decided server-side

**Comment by dydzio0614** [comment]
Oh and btw idk if it qualifies but poison effect from wyvern monarch may not work as in H3, to be checked if HP per turn mechanic is same as in H3, effect persistence after expiration and current HP being lowered together with max HP if greater than new max HP

**Comment by dydzio0614** [comment]
https://bugs.vcmi.eu/view.php?id=2614 one more bug that I remember was never fixed

**Comment by rilian-la-te** [comment]
Is control shortcut exist in vanilla h3 to bypass shooting (or harpy return)?

**Comment by IvanSavenko** [comment]
As far as I know - no.
However I assume there will be mod for 1.2 with this option by Nordsoft

**Comment by kambala-decapitator** [comment]
> 0001791: Orb of vulnerability mechanics in vcmi is different from H3

I have also discovered that in VCMI the Orb doesn't negate Magic Resistance of Crystal Dragons (and probably same goes to [Battle] Dwarves). Primary sources of info (https://heroes.thelazy.net/index.php/Orb_of_Vulnerability and FizMig) don't give any details on this thought, but [this post](http://heroescommunity.com/viewthread.php3?TID=21273&PID=1504564#focus) says that it should be negated.

**Comment by rilian-la-te** [comment]
950 from Mantis is fixed by battlefield actions submod.

**Comment by kambala-decapitator** [comment]
to add to the Orb of Vulnerability, the following shouldn't be affected by the Orb according to the document linked by @DjWarmonger on slack today (page 42):

- All living creatures are immune to Destroy Undead and Animate Dead.
- Unicorns: 20% probability to protect adjacent allies from all hostile hero cast spells except Firewall, Quicksand and Land Mine. Also grants immunity (with the same probability) to Thunderbird’s Lightning Strike, (War) Unicorn's Blind, Zombie's Disease, Enchanter’s Slow and Weakness, Black (Dread) Knight and Mummy’s Curse and Dragon (Serpent) Fly's Weakness.
- Efreets / Firebirds / Fire Elementals: Efreet Sultan's Fire Shield
- Troglodytes: Blind and Petrify
- Azure Dragons: AD's Fear
- All unliving creatures (Gargoyles, Golems and Elementals) are immune to Destroy Undead, Animate Dead, Resurrection, and Sacrifice. They are also immune to the regeneration ability of the Elixir of Life. Of creature abilities they are immune to Archangel’s Resurrection, Ghost Dragon's Aging, Mighty Gorgon's Death Stare, Pit Lord's Demon Summoning, Power Lich's Death Cloud (only adjacent splash damage), Vampire Lord's Life Drain, Wyvern Monarch's Poison.
- Golems / Elementals: (Master Genie’s) Mirth, Sorrow and Azure Dragon's Fear.
- Undead Creatures (Necropolis + Mummy) are immune to some Mind spells (Blind, Mirth, Sorrow, Fear). They are immune to Resurrection, Sacrifice, Death Ripple, Bless and Curse, and to the regeneration ability of the Elixir of Life. Of creature abilities they are immune to Archangel’s Resurrection, Azure Dragon's Fear, Black (Dread) Knight And Mummy’s Curse, Enchanter’s Bless, Ghost Dragon's Aging, Master Genie’s Bless and Mirth, Mighty Gorgon's Death Stare, Pit Lord's Demon Summoning, Power Lich's Death Cloud (only adjacent splash damage), Vampire Lord's Life Drain, (War) Unicorn's Blind, Wyvern Monarch's Poison and Zombie's Disease.
- war machines: Mind spells, Bloodlust, Implosion, Clone, Teleport.
- All non shooters are immune to Forgetfulness and Precision.
